src/Microsoft.Diagnostics.NETCore.Client/ReversedServer/ReversedDiagnosticsServer.cs

```diff
@@ -395,24 +465,63 @@ private bool TestStream(Stream stream)
             return false;
         }
 
-        private void RegisterHandler(Guid runtimeId, StreamHandler handler)
+        private void ProvideEndpointInfo(in IpcEndpointInfo endpointInfo)
         {
-            lock (_lock)
+            lock (_newEndpointInfoLock)
             {
-                if (!_cachedStreams.TryGetValue(runtimeId, out Stream stream))
+                IpcEndpointInfo? endpointInfoNullable = endpointInfo;
+                // Provide the endpoint info to the first handler that will accept it.
+                for (int i = 0; i < _newEndpointInfoHandlers.Count && endpointInfoNullable.HasValue; i++)
                 {
-                    throw new InvalidOperationException($"Runtime instance with identifier '{runtimeId}' is not registered.");
+                    EndpointInfoHandler handler = _newEndpointInfoHandlers[i];
+                    // Handler will return true if it has completed
+                    if (handler(ref endpointInfoNullable))
+                    {
+                        _newEndpointInfoHandlers.RemoveAt(i);
+                        i--;
+                    }
                 }
 
-                _handlers.Add(handler);
+                // If the endpoint info was not consumed, add it to the list
+                if (endpointInfoNullable.HasValue)
+                {
+                    _newEndpointInfos.Add(endpointInfo);
+                }
+            }
+        }
 
-                if (stream != null)
+        private void RegisterEndpointInfoHandler(EndpointInfoHandler handler)
+        {
+            lock (_newEndpointInfoLock)
+            {
+                // Attempt to accept an endpoint info
+                for (int i = 0; i < _newEndpointInfos.Count && null != handler; i++)
                 {
-                    RunStreamHandlers(runtimeId, stream);
+                    IpcEndpointInfo? endpointInfoNullable = _newEndpointInfos[i];
+                    // Handler will return true if it has completed
+                    if (handler(ref endpointInfoNullable))
+                    {
+                        handler = null;
+                    }
+
+                    // Handler will set endpoint info parameter to null when it is consumed.
+                    if (!endpointInfoNullable.HasValue)
+                    {
+                        _newEndpointInfos.RemoveAt(i);
+                        i--;
```

Comment:
So based on the code, if I make 10 calls to AcceptAsync, and then 1 connection is established, each one of those calls will trigger completion correct? Similarly what will happen here if I have 10 connections and I do 1 AcceptAsync call?

Reply:
Just for clarity for others:
10 AcceptAsync calls are awating and 1 connection comes in: the first AcceptAsync call is satisfied with the information; the other 9 are still pending.
10 connections and then 1 AcceptAsync call comes in: that call only accepts the first connection; thus 9 are still waiting to be consumed. Subsequent calls to AcceptAsync are needed to get the 9 remaining connections.